File: scripts/import-finance.py
#!/usr/bin/env python3
"""
TaskFlow Finance — One-time CSV Import Script (Python version)

Reads CSV exports from Stripe (2 accounts), Zoho Payments, and Brex,
normalises them, and inserts into the finance_payments table in Supabase.

Usage:
    python3 scripts/import-finance.py

The script is idempotent — it checks source + source_id to skip duplicates.
"""

import csv
import json
import re
import sys
import urllib.request
import urllib.error
import urllib.parse
from datetime import datetime

# ── Config ──────────────────────────────────────────────────────────
SUPABASE_URL = "https://tnkmxmlgdhlgehlrbxuf.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InRua214bWxnZGhsZ2VobHJieHVmIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3MjE0MjQzMiwiZXhwIjoyMDg3NzE4NDMyfQ.KRgDU-OCfx4inIU_tl0W24XfiojjFQ7K8jTYYyc0qlI"
USER_ID = "78bd1255-f05a-436b-abbd-f8c281d30210"

REST_ENDPOINT = SUPABASE_URL + "/rest/v1/finance_payments"
CSV_DIR = "/Users/timjarvis/Documents/TaskFlow Data/Finance"

# ── Brex: names to EXCLUDE (internal transfers, expenses, pass-through) ──
BREX_EXCLUDE = [
    "Client Advertising Account",
    "Cash Reserve",
    "Brex Card",
    "STRIPE",
    "TIM JARVIS ONLIN",
    "Film&Content LLC",
    "Lee Brack",
    "Reservoir Limited",
    "Dreyfus Government",
    "Brex",
    "PAYPAL",
    "IMPACT RADIUS",
    "Treasury",
    "Stripe - Do Not Use",
    "Nightwolf Productions",  # ACH return, not a real payment
    "CITYCOAST TRUST",
]


def brex_is_excluded(to_from):
    tf = to_from.lower()
    for name in BREX_EXCLUDE:
        if name.lower() in tf:
            return True
    return False


# ── Parse helpers ───────────────────────────────────────────────────
def parse_date(s):
    if not s:
        return None
    s = s.strip()

    # MM/DD/YYYY (Brex)
    m = re.match(r"^(\d{2})/(\d{2})/(\d{4})$", s)
    if m:
        return f"{m.group(3)}-{m.group(1)}-{m.group(2)}"

    # YYYY-MM-DD (already ISO)
    m = re.match(r"^(\d{4})-(\d{2})-(\d{2})", s)
    if m:
        return f"{m.group(1)}-{m.group(2)}-{m.group(3)}"

    # MM-DD-YYYY (Mercury)
    m = re.match(r"^(\d{2})-(\d{2})-(\d{4})", s)
    if m:
        return f"{m.group(3)}-{m.group(1)}-{m.group(2)}"

    # "Jan 28, 2026, 04:04 PM" (Zoho) — try common datetime formats
    for fmt in (
        "%b %d, %Y, %I:%M %p",
        "%b %d, %Y",
        "%B %d, %Y, %I:%M %p",
        "%B %d, %Y",
        "%m/%d/%Y %I:%M %p",
        "%Y-%m-%dT%H:%M:%S",
    ):
        try:
            dt = datetime.strptime(s, fmt)
            return dt.strftime("%Y-%m-%d")
        except ValueError:
            continue

    return None


def num(v):
    if not v:
        return 0.0
    cleaned = re.sub(r"[,$]", "", str(v))
    try:
        return float(cleaned)
    except ValueError:
        return 0.0


# ── Read & parse CSV ────────────────────────────────────────────────
def read_csv(filename):
    import os
    fp = os.path.join(CSV_DIR, filename)
    if not os.path.exists(fp):
        print(f"File not found: {fp}", file=sys.stderr)
        return []
    rows = []
    with open(fp, newline="", encoding="utf-8-sig") as f:
        reader = csv.DictReader(f)
        for row in reader:
            rows.append(row)
    return rows


# ── Transform functions per source ──────────────────────────────────

def transform_stripe1(rows):
    out = []
    for r in rows:
        if (r.get("Status") or "").strip() != "Paid":
            continue
        amt = num(r.get("Amount"))
        if amt <= 0:
            continue
        out.append({
            "user_id":     USER_ID,
            "date":        parse_date(r.get("Created date (UTC)")),
            "amount":      amt,
            "fee":         num(r.get("Fee")),
            "net":         amt - num(r.get("Fee")),
            "source":      "stripe",
            "source_id":   (r.get("id") or "").strip(),
            "payer_email": (r.get("Customer Email") or "").strip().lower(),
            "payer_name":  (r.get("Customer Description") or r.get("name (metadata)") or "").strip(),
            "description": (r.get("Description") or "").strip(),
            "category":    "",
            "client_id":   None,
            "campaign_id": None,
            "end_client":  "",
            "notes":       "",
            "status":      "unmatched",
        })
    return out


def transform_stripe2(rows):
    out = []
    for r in rows:
        if (r.get("Status") or "").strip() != "Paid":
            continue
        amt = num(r.get("Amount"))
        if amt <= 0:
            continue
        out.append({
            "user_id":     USER_ID,
            "date":        parse_date(r.get("Created date (UTC)")),
            "amount":      amt,
            "fee":         num(r.get("Fee")),
            "net":         amt - num(r.get("Fee")),
            "source":      "stripe2",
            "source_id":   (r.get("id") or "").strip(),
            "payer_email": (r.get("Customer Email") or "").strip().lower(),
            "payer_name":  (r.get("Customer Description") or r.get("name (metadata)") or "").strip(),
            "description": (r.get("Description") or "").strip(),
            "category":    "",
            "client_id":   None,
            "campaign_id": None,
            "end_client":  "",
            "notes":       "",
            "status":      "unmatched",
        })
    return out


def transform_zoho(rows):
    out = []
    for r in rows:
        if (r.get("Status") or "").strip() != "succeeded":
            continue
        amt = num(r.get("Amount"))
        if amt <= 0:
            continue
        out.append({
            "user_id":     USER_ID,
            "date":        parse_date(r.get("Date")),
            "amount":      amt,
            "fee":         num(r.get("FeeAmount")),
            "net":         amt - num(r.get("FeeAmount")),
            "source":      "zoho",
            "source_id":   (r.get("PaymentID") or "").strip(),
            "payer_email": (r.get("ReceiptEmail") or "").strip().lower(),
            "payer_name":  (r.get("CardHolderName") or "").strip(),
            "description": (r.get("Description") or "").strip(),
            "category":    "",
            "client_id":   None,
            "campaign_id": None,
            "end_client":  "",
            "notes":       "",
            "status":      "unmatched",
        })
    return out


def transform_brex(rows):
    out = []
    for r in rows:
        if (r.get("Status") or "").strip() != "Complete":
            continue
        amt = num(r.get("Amount"))
        if amt <= 0:
            continue
        to_from = (r.get("To/From") or "").strip()
        if brex_is_excluded(to_from):
            continue
        method = (r.get("Method") or "").strip()
        if method not in ("ACH", "Wire", "ACH return"):
            continue
        out.append({
            "user_id":     USER_ID,
            "date":        parse_date(r.get("Date")),
            "amount":      amt,
            "fee":         0,
            "net":         amt,
            "source":      "brex",
            "source_id":   "",
            "payer_email": "",
            "payer_name":  to_from,
            "description": (r.get("Memo") or r.get("External Memo") or "").strip(),
            "category":    "",
            "client_id":   None,
            "campaign_id": None,
            "end_client":  "",
            "notes":       f"Method: {method}",
            "status":      "unmatched",
        })
    return out


# ── Supabase REST helpers ───────────────────────────────────────────

def supabase_get(params):
    """GET request to the finance_payments REST endpoint with query params."""
    qs = urllib.parse.urlencode(params)
    url = REST_ENDPOINT + "?" + qs
    req = urllib.request.Request(url, method="GET")
    req.add_header("apikey", SUPABASE_KEY)
    req.add_header("Authorization", f"Bearer {SUPABASE_KEY}")
    req.add_header("Content-Type", "application/json")
    try:
        with urllib.request.urlopen(req) as resp:
            body = resp.read().decode("utf-8")
            return json.loads(body), None
    except urllib.error.HTTPError as e:
        err_body = e.read().decode("utf-8")
        return None, f"HTTP {e.code}: {err_body}"


def supabase_post(records):
    """POST (insert) a batch of records into finance_payments."""
    data = json.dumps(records).encode("utf-8")
    req = urllib.request.Request(REST_ENDPOINT, data=data, method="POST")
    req.add_header("apikey", SUPABASE_KEY)
    req.add_header("Authorization", f"Bearer {SUPABASE_KEY}")
    req.add_header("Content-Type", "application/json")
    req.add_header("Prefer", "return=minimal")
    try:
        with urllib.request.urlopen(req) as resp:
            return None  # success
    except urllib.error.HTTPError as e:
        err_body = e.read().decode("utf-8")
        return f"HTTP {e.code}: {err_body}"


# ── Main ────────────────────────────────────────────────────────────

def main():
    print("TaskFlow Finance Import")
    print("=======================\n")

    # 1. Read & transform
    print(f"Reading CSVs from: {CSV_DIR}")

    stripe1_rows = read_csv("stripe_payments.csv")
    stripe2_rows = read_csv("stripe_account2_payments.csv")
    zoho_rows    = read_csv("zoho_payments.csv")
    brex_rows    = read_csv("brex_payments.csv")

    stripe1 = transform_stripe1(stripe1_rows)
    stripe2 = transform_stripe2(stripe2_rows)
    zoho    = transform_zoho(zoho_rows)
    brex    = transform_brex(brex_rows)

    print(f"Stripe Account 1: {len(stripe1)} records")
    print(f"Stripe Account 2: {len(stripe2)} records")
    print(f"Zoho Payments:    {len(zoho)} records")
    print(f"Brex Direct:      {len(brex)} records")

    all_records = stripe1 + stripe2 + zoho + brex
    print(f"Total to import:  {len(all_records)} records\n")

    if not all_records:
        print("Nothing to import.")
        return

    # 2. Check for existing records to avoid duplicates
    print("Checking for existing records...")
    existing, err = supabase_get({
        "select": "source,source_id",
        "user_id": f"eq.{USER_ID}",
    })

    if err:
        print(f"Error checking existing: {err}", file=sys.stderr)
        return

    existing_set = set()
    for r in (existing or []):
        if r.get("source_id"):
            existing_set.add(r["source"] + ":" + r["source_id"])
    print(f"Existing records: {len(existing_set)}")

    # Filter out duplicates (only for records with source_id)
    to_insert = []
    for r in all_records:
        if not r["source_id"]:
            to_insert.append(r)  # Brex has no source_id, always insert
        elif (r["source"] + ":" + r["source_id"]) not in existing_set:
            to_insert.append(r)

    print(f"After dedup:      {len(to_insert)} new records\n")

    if not to_insert:
        print("All records already imported. Nothing to do.")
        return

    # 3. Insert in batches of 50
    BATCH = 50
    inserted = 0
    errors = 0

    for i in range(0, len(to_insert), BATCH):
        batch = to_insert[i : i + BATCH]
        err = supabase_post(batch)
        if err:
            batch_num = (i // BATCH) + 1
            print(f"Insert error (batch {batch_num}): {err}", file=sys.stderr)
            errors += len(batch)
        else:
            inserted += len(batch)
            sys.stdout.write(f"\rInserted: {inserted}/{len(to_insert)}")
            sys.stdout.flush()

    print("\n\nImport complete!")
    print(f"  Inserted: {inserted}")
    print(f"  Errors:   {errors}")

    # 4. Summary by source
    print("\n--- Revenue Summary ---")
    sources = {"stripe": 0.0, "stripe2": 0.0, "zoho": 0.0, "brex": 0.0}
    for r in to_insert:
        sources[r["source"]] += r["amount"]
    for s, total in sources.items():
        print(f"  {s}: ${total:.2f}")
    grand_total = sum(sources.values())
    print(f"  Total: ${grand_total:.2f}")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"Fatal error: {e}", file=sys.stderr)
        sys.exit(1)

File: scripts/import-finance.js
#!/usr/bin/env node
/**
 * TaskFlow Finance — One-time CSV Import Script
 *
 * Reads CSV exports from Stripe (2 accounts), Zoho Payments, and Brex,
 * normalises them, and inserts into the finance_payments table in Supabase.
 *
 * Usage:
 *   1. npm install @supabase/supabase-js csv-parse
 *   2. Set SUPABASE_URL, SUPABASE_KEY, and USER_ID below (or via env)
 *   3. node scripts/import-finance.js
 *
 * The script is idempotent — it checks source + source_id to skip duplicates.
 */

var { createClient } = require('@supabase/supabase-js');
var fs = require('fs');
var path = require('path');
var { parse } = require('csv-parse/sync');

// ── Config ──────────────────────────────────────────────────────────
var SUPABASE_URL = process.env.SUPABASE_URL || 'https://tnkmxmlgdhlgehlrbxuf.supabase.co';
var SUPABASE_KEY = process.env.SUPABASE_SERVICE_KEY || '';  // USE SERVICE ROLE KEY (not anon) to bypass RLS
var USER_ID      = process.env.USER_ID || '';               // Your auth.users UUID

if (!SUPABASE_KEY) { console.error('Set SUPABASE_SERVICE_KEY env var'); process.exit(1); }
if (!USER_ID)      { console.error('Set USER_ID env var'); process.exit(1); }

var sb = createClient(SUPABASE_URL, SUPABASE_KEY);

var CSV_DIR = path.join(__dirname, '..', '..', 'TaskFlow Data', 'Finance');

// ── Brex: names to EXCLUDE (internal transfers, expenses, pass-through) ──
var BREX_EXCLUDE = [
  'Client Advertising Account',
  'Cash Reserve',
  'Brex Card',
  'STRIPE',
  'TIM JARVIS ONLIN',
  'Film&Content LLC',
  'Lee Brack',
  'Reservoir Limited',
  'Dreyfus Government',
  'Brex',
  'PAYPAL',
  'IMPACT RADIUS',
  'Treasury',
  'Stripe - Do Not Use',
  'Nightwolf Productions',  // ACH return, not a real payment
  'CITYCOAST TRUST'
];

function brexIsExcluded(toFrom) {
  var tf = toFrom.toLowerCase();
  for (var i = 0; i < BREX_EXCLUDE.length; i++) {
    if (tf.indexOf(BREX_EXCLUDE[i].toLowerCase()) !== -1) return true;
  }
  return false;
}

// ── Parse helpers ───────────────────────────────────────────────────
function parseDate(str) {
  if (!str) return null;
  str = str.trim();

  // MM/DD/YYYY (Brex)
  var m1 = str.match(/^(\d{2})\/(\d{2})\/(\d{4})$/);
  if (m1) return m1[3] + '-' + m1[1] + '-' + m1[2];

  // YYYY-MM-DD (already ISO)
  var m2 = str.match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (m2) return m2[1] + '-' + m2[2] + '-' + m2[3];

  // MM-DD-YYYY (Mercury)
  var m3 = str.match(/^(\d{2})-(\d{2})-(\d{4})/);
  if (m3) return m3[3] + '-' + m3[1] + '-' + m3[2];

  // "Jan 28, 2026, 04:04 PM" (Zoho)
  var d = new Date(str);
  if (!isNaN(d.getTime())) {
    var y = d.getFullYear();
    var mo = String(d.getMonth() + 1).padStart(2, '0');
    var da = String(d.getDate()).padStart(2, '0');
    return y + '-' + mo + '-' + da;
  }
  return null;
}

function num(v) {
  if (!v) return 0;
  return parseFloat(String(v).replace(/[,$]/g, '')) || 0;
}

// ── Read & parse CSV ────────────────────────────────────────────────
function readCSV(filename) {
  var fp = path.join(CSV_DIR, filename);
  if (!fs.existsSync(fp)) { console.error('File not found:', fp); return []; }
  var raw = fs.readFileSync(fp, 'utf-8');
  return parse(raw, { columns: true, skip_empty_lines: true, relax_column_count: true });
}

// ── Transform functions per source ──────────────────────────────────

function transformStripe1(rows) {
  var out = [];
  for (var i = 0; i < rows.length; i++) {
    var r = rows[i];
    if ((r['Status'] || '').trim() !== 'Paid') continue;
    var amt = num(r['Amount']);
    if (amt <= 0) continue;
    out.push({
      user_id:     USER_ID,
      date:        parseDate(r['Created date (UTC)']),
      amount:      amt,
      fee:         num(r['Fee']),
      net:         amt - num(r['Fee']),
      source:      'stripe',
      source_id:   (r['id'] || '').trim(),
      payer_email: (r['Customer Email'] || '').trim().toLowerCase(),
      payer_name:  (r['Customer Description'] || r['name (metadata)'] || '').trim(),
      description: (r['Description'] || '').trim(),
      category:    '',
      client_id:   null,
      campaign_id: null,
      end_client:  '',
      notes:       '',
      status:      'unmatched'
    });
  }
  return out;
}

function transformStripe2(rows) {
  var out = [];
  for (var i = 0; i < rows.length; i++) {
    var r = rows[i];
    if ((r['Status'] || '').trim() !== 'Paid') continue;
    var amt = num(r['Amount']);
    if (amt <= 0) continue;
    out.push({
      user_id:     USER_ID,
      date:        parseDate(r['Created date (UTC)']),
      amount:      amt,
      fee:         num(r['Fee']),
      net:         amt - num(r['Fee']),
      source:      'stripe2',
      source_id:   (r['id'] || '').trim(),
      payer_email: (r['Customer Email'] || '').trim().toLowerCase(),
      payer_name:  (r['Customer Description'] || r['name (metadata)'] || '').trim(),
      description: (r['Description'] || '').trim(),
      category:    '',
      client_id:   null,
      campaign_id: null,
      end_client:  '',
      notes:       '',
      status:      'unmatched'
    });
  }
  return out;
}

function transformZoho(rows) {
  var out = [];
  for (var i = 0; i < rows.length; i++) {
    var r = rows[i];
    if ((r['Status'] || '').trim() !== 'succeeded') continue;
    var amt = num(r['Amount']);
    if (amt <= 0) continue;
    out.push({
      user_id:     USER_ID,
      date:        parseDate(r['Date']),
      amount:      amt,
      fee:         num(r['FeeAmount']),
      net:         amt - num(r['FeeAmount']),
      source:      'zoho',
      source_id:   (r['PaymentID'] || '').trim(),
      payer_email: (r['ReceiptEmail'] || '').trim().toLowerCase(),
      payer_name:  (r['CardHolderName'] || '').trim(),
      description: (r['Description'] || '').trim(),
      category:    '',
      client_id:   null,
      campaign_id: null,
      end_client:  '',
      notes:       '',
      status:      'unmatched'
    });
  }
  return out;
}

function transformBrex(rows) {
  var out = [];
  for (var i = 0; i < rows.length; i++) {
    var r = rows[i];
    if ((r['Status'] || '').trim() !== 'Complete') continue;
    var amt = num(r['Amount']);
    if (amt <= 0) continue;  // only incoming (positive)
    var toFrom = (r['To/From'] || '').trim();
    if (brexIsExcluded(toFrom)) continue;
    var method = (r['Method'] || '').trim();
    if (method !== 'ACH' && method !== 'Wire' && method !== 'ACH return') continue;
    out.push({
      user_id:     USER_ID,
      date:        parseDate(r['Date']),
      amount:      amt,
      fee:         0,
      net:         amt,
      source:      'brex',
      source_id:   '',
      payer_email: '',
      payer_name:  toFrom,
      description: (r['Memo'] || r['External Memo'] || '').trim(),
      category:    '',
      client_id:   null,
      campaign_id: null,
      end_client:  '',
      notes:       'Method: ' + method,
      status:      'unmatched'
    });
  }
  return out;
}

// ── Main ────────────────────────────────────────────────────────────
async function main() {
  console.log('TaskFlow Finance Import');
  console.log('=======================\n');

  // 1. Read & transform
  console.log('Reading CSVs from:', CSV_DIR);

  var stripe1Rows = readCSV('stripe_payments.csv');
  var stripe2Rows = readCSV('stripe_account2_payments.csv');
  var zohoRows    = readCSV('zoho_payments.csv');
  var brexRows    = readCSV('brex_payments.csv');

  var stripe1 = transformStripe1(stripe1Rows);
  var stripe2 = transformStripe2(stripe2Rows);
  var zoho    = transformZoho(zohoRows);
  var brex    = transformBrex(brexRows);

  console.log('Stripe Account 1:', stripe1.length, 'records');
  console.log('Stripe Account 2:', stripe2.length, 'records');
  console.log('Zoho Payments:   ', zoho.length, 'records');
  console.log('Brex Direct:     ', brex.length, 'records');

  var all = stripe1.concat(stripe2).concat(zoho).concat(brex);
  console.log('Total to import: ', all.length, 'records\n');

  if (all.length === 0) {
    console.log('Nothing to import.');
    return;
  }

  // 2. Check for existing records to avoid duplicates
  console.log('Checking for existing records...');
  var { data: existing, error: existErr } = await sb
    .from('finance_payments')
    .select('source, source_id')
    .eq('user_id', USER_ID);

  if (existErr) {
    console.error('Error checking existing:', existErr.message);
    return;
  }

  var existingSet = new Set();
  (existing || []).forEach(function(r) {
    if (r.source_id) existingSet.add(r.source + ':' + r.source_id);
  });
  console.log('Existing records:', existingSet.size);

  // Filter out duplicates (only for records with source_id)
  var toInsert = all.filter(function(r) {
    if (!r.source_id) return true;  // Brex has no source_id, always insert
    return !existingSet.has(r.source + ':' + r.source_id);
  });

  console.log('After dedup:     ', toInsert.length, 'new records\n');

  if (toInsert.length === 0) {
    console.log('All records already imported. Nothing to do.');
    return;
  }

  // 3. Insert in batches of 50
  var BATCH = 50;
  var inserted = 0;
  var errors = 0;

  for (var i = 0; i < toInsert.length; i += BATCH) {
    var batch = toInsert.slice(i, i + BATCH);
    var { error: insErr } = await sb.from('finance_payments').insert(batch);
    if (insErr) {
      console.error('Insert error (batch', Math.floor(i / BATCH) + 1, '):', insErr.message);
      errors += batch.length;
    } else {
      inserted += batch.length;
      process.stdout.write('Inserted: ' + inserted + '/' + toInsert.length + '\r');
    }
  }

  console.log('\n\nImport complete!');
  console.log('  Inserted:', inserted);
  console.log('  Errors:  ', errors);

  // 4. Summary by source
  console.log('\n--- Revenue Summary ---');
  var sources = { stripe: 0, stripe2: 0, zoho: 0, brex: 0 };
  toInsert.forEach(function(r) { sources[r.source] += r.amount; });
  Object.keys(sources).forEach(function(s) {
    console.log('  ' + s + ': $' + sources[s].toFixed(2));
  });
  console.log('  Total: $' + Object.values(sources).reduce(function(a, b) { return a + b; }, 0).toFixed(2));
}

main().catch(function(err) {
  console.error('Fatal error:', err);
  process.exit(1);
});
